src/Hosting/TestHost/test/TestClientTests.cs

```diff
@@ -384,6 +384,53 @@ public async Task ClientStreaming_ResponseCompletesWithoutReadingRequest()
             await writeTask;
         }
 
+        [Fact]
+        public async Task ClientStreaming_ResponseCompletesWithoutResponseBodyWrite()
+        {
+            // Arrange
+            var requestStreamTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            RequestDelegate appDelegate = ctx =>
+            {
+                ctx.Response.Headers["test-header"] = "true";
+                return Task.CompletedTask;
+            };
+
+            Stream requestStream = null;
+
+            var builder = new WebHostBuilder().Configure(app => app.Run(appDelegate));
+            var server = new TestServer(builder);
+            var client = server.CreateClient();
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, "http://localhost:12345");
+            httpRequest.Version = new Version(2, 0);
+            httpRequest.Content = new PushContent(async stream =>
+            {
+                requestStream = stream;
+                await requestStreamTcs.Task;
+            });
+
+            // Act
+            var response = await client.SendAsync(httpRequest, HttpCompletionOption.ResponseHeadersRead).WithTimeout();
+
+            var responseContent = await response.Content.ReadAsStreamAsync().WithTimeout();
+
+            // Assert
+            response.EnsureSuccessStatusCode();
+            Assert.Equal("true", response.Headers.GetValues("test-header").Single());
+
+            // Read response
+            byte[] buffer = new byte[1024];
+            var length = await responseContent.ReadAsync(buffer).AsTask().WithTimeout();
+            Assert.Equal(0, length);
+
+            // Writing to request stream will fail because server is complete
+            await Assert.ThrowsAnyAsync<Exception>(() => requestStream.WriteAsync(buffer).AsTask());
```

Comment:
What exception does this throw? Kestrel has a 5-second drain timeout where the client can still send request body data after the response is completed, so this doesn't exactly match Kestrel's behavior.

Reply:
TestServer does not try to preserve client observable behavior, only app visible behavior. At this point the app delegate has returned so there's no need to keep the request stream open.

Kestrel also has that drain timeout to avoid closing connections and there are no connections here to be concerned with.

Reply:
> What exception does this throw?

I believe it is an IOException. The request stream throws it once the request is complete on the server.